lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs

```diff
@@ -1185,23 +1247,42 @@ _assignScriptRedeemers (toAlonzoPParams -> pparams) ti resolveInput redeemers tx
          in
             Ledger.UTxO (Map.fromList utxo)
 
+    utxoFromBabbageTx
+        :: BabbageTx
+        -> Ledger.UTxO (Cardano.ShelleyLedgerEra Cardano.BabbageEra)
+    utxoFromBabbageTx babbageTx =
+        let
+            inputs = Babbage.inputs (Alonzo.body babbageTx)
+            utxo = flip mapMaybe (F.toList inputs) $ \i -> do
+                (o, dt) <- resolveInput (fromShelleyTxIn i)
+                -- NOTE: 'toAlonzoTxOut' only partially represent the information
+                -- because the wallet internal types aren't capturing datum at
+                -- the moment. It is _okay_ in this context because the
+                -- resulting UTxO is only used by 'evaluateTransactionExecutionUnits'
+                -- to lookup addresses corresponding to inputs.
+                pure (i, toBabbageTxOut o dt)
+         in
+            Ledger.UTxO (Map.fromList utxo)
+
+
     -- | Evaluate execution units of each script/redeemer in the transaction.
     -- This may fail for each script.
-    evaluateExecutionUnits
+    evaluateExecutionUnitsAlonzo
         :: Map Alonzo.RdmrPtr Redeemer
         -> AlonzoTx
         -> Either ErrAssignRedeemers
             (Map Alonzo.RdmrPtr (Either ErrAssignRedeemers Alonzo.ExUnits))
-    evaluateExecutionUnits indexedRedeemers alonzoTx = do
+    evaluateExecutionUnitsAlonzo indexedRedeemers alonzoTx = do
         let utxo = utxoFromAlonzoTx alonzoTx
-        let costs = toCostModelsAsArray (Alonzo._costmdls pparams)
+        let pparams' = Cardano.toLedgerPParams Cardano.ShelleyBasedEraAlonzo pparams
+        let costs = toCostModelsAsArray (Alonzo.unCostModels $ Alonzo._costmdls pparams')
         let systemStart = getSystemStart ti
 
-        epochInfo <- hoistEpochInfo (left ErrAssignRedeemersPastHorizon . runIdentity . runExceptT)
+        epochInfo <- hoistEpochInfo (left (T.pack . show . ErrAssignRedeemersPastHorizon) . runIdentity . runExceptT)
             <$> left ErrAssignRedeemersPastHorizon (toEpochInfo ti)
 
-        res <- evaluateTransactionExecutionUnits
-                pparams
+        let res = evaluateTransactionExecutionUnits
```

Comment:
🧐 I think the `ErrAssignRedeemersPastHorizon` gets lost here. (`toEpochInfo` above won't throw it, which is a flaw of `TimeInterpreter`, but still, whether we get a PastHorizon or not would only be known when `evaluateTransactionExecutionUnits` actually tries to convert the validity interval.)

Will try to fix.

Reply:
Ah, it seems we no longer get the actual `PastHorizonException` from the ledger 😬 https://github.com/input-output-hk/cardano-ledger/blob/e290bf8d0ea272a51e9acd10adc96b4e12e00d37/eras/alonzo/impl/src/Cardano/Ledger/Alonzo/TxInfo.hs#L106

Reply:
We probably should remove `ErrAssignRedeemersPastHorizon` then, and expose each constructor of `TranslationError` separately in the our API.

For now I removed `ErrAssignRedeemersPastHorizon`: 3c62f15f3b7879b4e0460ab20647b0280e0a9d4c

Reply:
I pushed 896c0680766b77fc761bdf0313d8ca25977bcbdb to minimise the API breakage from this PR and keep using the `past_horizon` error code. cc @jonathanknowles does it sound reasonable?